metrics/opencensus_exporter.go

```diff
@@ -36,3 +55,82 @@ func newOpenCensusExporter(config *metricsConfig, logger *zap.SugaredLogger) (vi
 	view.RegisterExporter(e)
 	return e, nil
 }
+
+func defaultResourceDescriptor(ctx context.Context) (*resource.Resource, error) {
+	if r, ok := metricskey.FromContext(ctx); ok {
+		return r, nil
+	}
+	m := tag.FromContext(ctx)
+
+	if _, ok := m.Value(tag.MustNewKey(metricskey.LabelRevisionName)); ok {
+		return &resource.Resource{
+			Type:   metricskey.ResourceTypeKnativeRevision,
+			Labels: extractLabelsFromMap(m, metricskey.KnativeRevisionLabels.UnsortedList()...),
```

Comment:
So, unfortunately, the  "WithResourceDetector" option only runs the ResourceDetector once at time of exporter creation, so it can't capture dynamic label values like revision_name and configuration_name.

This isn't necessarily breaking problem since the correct values will still be passed along as normal metrics tags, but it will be very strange to have the resource labels differ from the metrics tags.

I'm not sure if the static nature of the Resource is by design for the ocagent or just a code limitation... I can find out from the Otel folks.

Reply:
Thanks for pointing this out.

I'm guessing that the common case for OpenCensus is that each object reports statistics about itself, rather than for infrastructure which reports on behalf of other objects. Do you want to see about making an API change upstream for this case?

Looking at the code, it looks like `Exporter.resource` is used in the following places:

**Trace:**
https://github.com/census-ecosystem/opencensus-go-exporter-ocagent/blob/master/ocagent.go#L232
https://github.com/census-ecosystem/opencensus-go-exporter-ocagent/blob/master/ocagent.go#L507

**View:**
https://github.com/census-ecosystem/opencensus-go-exporter-ocagent/blob/master/ocagent.go#L268
https://github.com/census-ecosystem/opencensus-go-exporter-ocagent/blob/master/ocagent.go#L540


Unfortunately, I think the assumption of a single resource may be somewhat baked in to the exporter, because the same resource is used for all the metrics exported from a View.

I'm wondering whether we should maintain multiple exporters in multi-tenant cases like the activator and controllers. We could do this without neeeding to change the interface by keeping a map from Resource -> Exporter in this OpenCensus module.

Reply:
Digging around the OpenCensus world, each OpenCensus Metric does actually support having a unique Resource field that can supercede the batch's Resource: https://github.com/census-instrumentation/opencensus-proto/blob/master/gen-go/metrics/v1/metrics.pb.go#L108

The ocagent does convert each row in a view individually, so we could modify it to re-run the ResourceDetector on each view.Data row: https://github.com/census-ecosystem/opencensus-go-exporter-ocagent/blob/master/transform_stats_to_metrics.go#L36 

I prefer your idea of doing an exporter per unique knative_revision Resource though because:
1. Even if we fix the ocagent exporter, OpenCensus collector service & other exporters rely on batching too, so they'd all need to respect Metric's Resource over the batched Resource. I know Stackdriver exporter will, but not sure about the others.
2. I'm not super familiar with contexts but it seems like the ocagent exporter's context isn't shared with the Knative controller's and won't have the required metrics tags anyway? https://github.com/census-ecosystem/opencensus-go-exporter-ocagent/blob/master/ocagent.go#L130

My only concern with one exporter per Resource is it might be a lot of exporters, but I think it would just be one exporter per revision? Would you agree on this set of Resource labels?
- revision_name
- configuration_name
- namespace_name
- service_name

Assuming the GCP stuff is going to be ripped out and shouldn't be part of the OpenCensus knative_revision Resource model (which can be added back on the collector service side).